# Date Night Movie




- Code: 6.5 pts
- Comments: .5pts
- Markdown Documentation: .5pts


In this assignment we are going to use pandas to figure out - What's the best **date-night movie**?

This assignment is going to use

- Joining
- Groupby
- Sorting


In [135]:
import os
import pandas as pd

##### Read in the movie data: `pd.read_table`

In [136]:
def get_movie_data():
    
    '''
    Function to read in movie related data
    
    Parameters
    ----------
    None
    
    Returns
    -------
    users: pd.DataFrame containing user data
    ratings: pd.DataFrame containing rating data
    movies: pd.DataFrame containing movie data
    '''
    
    unames = ['user_id','gender','age','occupation','zip']
    users = pd.read_csv('users.dat' ,sep='::', header=None, names=unames, engine='python')
    
    rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
    ratings = pd.read_csv('ratings.dat',sep='::', header=None, names=rnames, engine='python')
    
    mnames = ['movie_id', 'title','genres']
    movies = pd.read_csv('movies.dat', sep='::', header=None, names=mnames, engine='python')

    return users, ratings, movies

In [137]:
users, ratings, movies = get_movie_data()

In [138]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

In [139]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [140]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

##### Clean up the `movies`

- Get the `year`
- Shorten the `title`


In [141]:
tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')
tmp.apply(lambda x:x[0] if len(x) > 0 else None)
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)

0    Toy Story
1         1995
dtype: object

In [142]:
movies['year'] = tmp[1]
movies['short_title'] = tmp[0]

In [143]:
movies.head()

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II

In [144]:
movies.shape,users.shape,ratings.shape

((3883, 5), (6040, 5), (1000209, 4))

In [145]:
movies.columns,ratings.columns

(Index(['movie_id', 'title', 'genres', 'year', 'short_title'], dtype='object'),
 Index(['user_id', 'movie_id', 'rating', 'timestamp'], dtype='object'))

##### Join the tables with `pd.merge` (1 pts)

In [146]:
movie_ratings = pd.merge(movies, ratings)
user_rating= pd.merge(users, ratings)

##### What's the highest rated movie? (1 pts))

In [147]:
movie_ratings['rating'].max()
c=movie_ratings.groupby('short_title').rating.count()
import numpy as np
np.argmax(c)
c.iloc[126:127]
#The highest movie is American beauty

short_title
American Beauty    3428
Name: rating, dtype: int64

##### What are some good rated movies for date night? (4.5 pts)

- Hint - highly rated movie by 
    - both partners (might be the same gender or not),
    - based on genre preferences,
    - age group can also be combined
    - Perhaps occupation? 
    - etc.
    
There is no single correct answer. Be sure to explain your reasoning behind your suggestion

In [340]:
l=list(user_rating['movie_id'].unique())



In [341]:
l.sort()

In [342]:
user_rating.groupby('movie_id').gender.value_counts(sort=True)

movie_id  gender
1         M         1486
          F          591
2         M          525
          F          176
3         M          342
                    ... 
3950      F           10
3951      M           23
          F           17
3952      M          283
          F          105
Name: gender, Length: 7152, dtype: int64

In [343]:
u=user_rating.groupby('movie_id').gender.value_counts().to_numpy()

In [347]:
M=[u[2*k] for k in range(int(len(u)/2))]
F=[u[2*k+1] for k in range(int(len(u)/2))]

In [348]:
h=[min(M[i],F[i]) for i in range(len(M))]

In [350]:
max(h),np.argmax(h)


(798, 2126)

In [351]:
M[2126]

1571

In [352]:
for i in l:
    if user_rating[user_rating['movie_id']==i].gender.value_counts()[0]==1571:
        print(i)

2396


In [353]:
movies[movies['movie_id']==2396]

movie_id                       title          genres  year  \
2327      2396  Shakespeare in Love (1998)  Comedy|Romance  1998   

              short_title  
2327  Shakespeare in Love

Snake eyes

In [355]:
user_rating[user_rating['movie_id']==2396].gender.value_counts()

M    1571
F     798
Name: gender, dtype: int64

## The best good rated movies for date night is Shakespeare in Love (1998) because it's rated by large number of both gender (1571 M&798 F)